File: Queues/QueuesWithoutLibrary.java
package Queues;

/**
 * Demonstrates queue operations using custom linked list implementation.
 * Provides O(1) enqueue and dequeue operations without using standard library.
 * Enqueues 5 integers and dequeues them to show FIFO ordering.
 */
public class QueuesWithoutLibrary {

    /**
     * Node class for linked list implementation.
     */
    private static class Node {
        int data;
        Node next;

        Node(int value) {
            data = value;
            next = null;
        }
    }

    /**
     * Queue implementation using linked list.
     */
    private static class CustomQueue {
        private Node front;
        private Node rear;
        private int count;

        public CustomQueue() {
            front = rear = null;
            count = 0;
        }

        /**
         * Add element to the rear of the queue. O(1) operation.
         */
        public void enqueue(int value) {
            Node newNode = new Node(value);
            if (isEmpty()) {
                front = rear = newNode;
            } else {
                rear.next = newNode;
                rear = newNode;
            }
            count++;
        }

        /**
         * Remove and return front element from the queue. O(1) operation.
         */
        public int dequeue() {
            if (isEmpty()) {
                System.err.println("Error: Attempting to dequeue from an empty queue.");
                return -1;
            }

            int value = front.data;
            front = front.next;

            if (front == null) {
                rear = null;
            }
            count--;
            return value;
        }

        /**
         * Check if queue is empty. O(1) operation.
         */
        public boolean isEmpty() {
            return front == null;
        }

        /**
         * Return front element without removing it. O(1) operation.
         */
        public int peek() {
            if (isEmpty()) {
                System.err.println("Error: Queue is empty.");
                return -1;
            }
            return front.data;
        }

        /**
         * Return number of elements in queue. O(1) operation.
         */
        public int size() {
            return count;
        }
    }

    public static void main(String[] args) {
        Runtime runtime = Runtime.getRuntime();
        runtime.gc();
        long memoryBefore = runtime.totalMemory() - runtime.freeMemory();
        long startTime = System.nanoTime();
        
        // Create queue instance
        CustomQueue myQueue = new CustomQueue();

        // Enqueue elements
        myQueue.enqueue(10);
        myQueue.enqueue(20);
        myQueue.enqueue(30);
        myQueue.enqueue(40);
        myQueue.enqueue(50);

        // Dequeue and print elements
        System.out.println("Elements in the Queue:");
        while (!myQueue.isEmpty()) {
            System.out.println("Dequeued Element: " + myQueue.dequeue());
        }
        
        long endTime = System.nanoTime();
        long memoryAfter = runtime.totalMemory() - runtime.freeMemory();
        long memoryUsed = memoryAfter - memoryBefore;
        
        // Performance statistics
        System.out.println("\n--- Performance Statistics ---");
        System.out.printf("Execution time: %.4f ms%n", (endTime - startTime) / 1_000_000.0);
        System.out.println("Memory usage: " + (memoryUsed > 0 ? memoryUsed : "negligible") + " bytes");
        System.out.println("\nTime Complexity:");
        System.out.println("  - Enqueue: O(1)");
        System.out.println("  - Dequeue: O(1)");
        System.out.println("  - Peek: O(1)");
        System.out.println("  - IsEmpty: O(1)");
    }
}

File: Queues/QueuesWithoutLibrary.cpp
#include <iostream>
#include <chrono>
#include <iomanip>

/**
 * Node structure for linked list implementation of queue.
 */
struct Node {
    int data;
    Node* next;

    Node(int value) : data(value), next(nullptr) {}
};

/**
 * Queue implementation using linked list (without STL).
 * Provides O(1) enqueue and dequeue operations.
 */
class Queue {
private:
    Node* front;
    Node* rear;
    size_t count;

public:
    Queue() : front(nullptr), rear(nullptr), count(0) {}

    ~Queue() {
        while (!isEmpty()) {
            dequeue();
        }
    }

    /**
     * Add element to the rear of the queue. O(1) operation.
     */
    void enqueue(int value) {
        Node* newNode = new Node(value);
        if (isEmpty()) {
            front = rear = newNode;
        } else {
            rear->next = newNode;
            rear = newNode;
        }
        count++;
    }

    /**
     * Remove and return front element from the queue. O(1) operation.
     */
    int dequeue() {
        if (isEmpty()) {
            std::cerr << "Error: Attempting to dequeue from an empty queue." << std::endl;
            return -1;
        }

        int value = front->data;
        Node* temp = front;
        front = front->next;
        delete temp;

        if (front == nullptr) {
            rear = nullptr;
        }
        count--;
        return value;
    }

    /**
     * Check if queue is empty. O(1) operation.
     */
    bool isEmpty() const {
        return front == nullptr;
    }

    /**
     * Return front element without removing it. O(1) operation.
     */
    int peek() const {
        if (isEmpty()) {
            std::cerr << "Error: Queue is empty." << std::endl;
            return -1;
        }
        return front->data;
    }

    /**
     * Return number of elements in queue. O(1) operation.
     */
    size_t size() const {
        return count;
    }
};

/**
 * Demonstrates queue operations using custom linked list implementation.
 * Enqueues 5 integers and dequeues them to show FIFO ordering.
 */
int main() {
    auto startTime = std::chrono::high_resolution_clock::now();
    
    // Create queue instance
    Queue myQueue;

    // Enqueue elements
    myQueue.enqueue(10);
    myQueue.enqueue(20);
    myQueue.enqueue(30);
    myQueue.enqueue(40);
    myQueue.enqueue(50);
    
    // Calculate memory (5 nodes + overhead)
    size_t memoryUsed = sizeof(Queue) + (sizeof(Node) * 5);

    // Dequeue and print elements
    std::cout << "Elements in the Queue:" << std::endl;
    while (!myQueue.isEmpty()) {
        std::cout << "Dequeued Element: " << myQueue.dequeue() << std::endl;
    }
    
    auto endTime = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime);
    
    // Performance statistics
    std::cout << "\n--- Performance Statistics ---" << std::endl;
    std::cout << std::fixed << std::setprecision(4);
    std::cout << "Execution time: " << duration.count() / 1000.0 << " ms" << std::endl;
    std::cout << "Memory usage: " << memoryUsed << " bytes" << std::endl;
    std::cout << "\nTime Complexity:" << std::endl;
    std::cout << "  - Enqueue: O(1)" << std::endl;
    std::cout << "  - Dequeue: O(1)" << std::endl;
    std::cout << "  - Peek: O(1)" << std::endl;
    std::cout << "  - IsEmpty: O(1)" << std::endl;

    return 0;
}
